implementations/rust/ockam/ockam_api/src/nodes/service/portals.rs

```diff
@@ -85,31 +93,74 @@ impl NodeManagerWorker {
         req: &Request<'_>,
         dec: &mut Decoder<'_>,
     ) -> Result<ResponseBuilder<InletStatus<'a>>> {
+        let manager = self.node_manager.clone();
         let mut node_manager = self.node_manager.write().await;
-        let CreateInlet {
-            bind_addr,
-            outlet_route,
-            alias,
-            check_credential,
-            ..
-        } = dec.decode()?;
-        let bind_addr = bind_addr.to_string();
+        let rid = req.id();
+        let req: CreateInlet = dec.decode()?;
 
-        let alias = alias.map(|a| a.0.into()).unwrap_or_else(random_alias);
+        let listen_addr = req.listen_addr().to_string();
+        let alias = req
+            .alias()
+            .map(|a| a.to_string())
+            .unwrap_or_else(random_alias);
 
-        info!("Handling request to create inlet portal");
+        info! {
+            from = %req.listen_addr(),
+            to = %req.outlet_addr(),
+            "Handling request to create inlet portal"
+        }
+
+        let (outer, rest) = {
+            let (sec1, rest) = node_manager
+                .connect(
+                    req.outlet_addr(),
+                    CredentialExchangeMode::Oneway,
+                    req.authorized(),
+                )
+                .await?;
+            if !sec1.is_empty() && rest.matches(0, &[Service::CODE.into(), Secure::CODE.into()]) {
+                let addr = sec1.clone().try_with(rest.iter().take(2))?;
+                let (sec2, _) = node_manager
+                    .connect(&addr, CredentialExchangeMode::Mutual, None)
+                    .await?;
+                (sec1, sec2.try_with(rest.iter().skip(2))?)
+            } else {
+                (MultiAddr::default(), sec1.try_with(&rest)?)
+            }
+        };
```

Comment:
I'm having a hard time understanding the context of various parts of the code. For example, in this block, it would be very helpful to have a comment explaining what's going to happen, why it's needed, etc., so anyone can understand a chunk of code without having to read it.

Reply:
Great, thanks toralf! I'll finish the review as soon as I can :+1: